src/bluetooth_linux.c

```diff
@@ -0,0 +1,673 @@
+					if (!strcmp(uuid, FIDO_SERVICE_REVISION_UUID))
+						service_revision_found = true;
+				}
+			} else {
+				sd_bus_message_skip(reply, "v");
+			}
+		}
+		sd_bus_message_exit_container(reply);
+	}
+	sd_bus_message_exit_container(reply);
+	if (!matches)
+	       return;
+
+	if (status_found)
+		newdev->paths.status = strdup(path);
+
+	if (control_point_found)
+		newdev->paths.control_point = strdup(path);
+
+	if (control_point_length_found)
+		newdev->paths.control_point_length = strdup(path);
+
+	if (service_revision_found)
+		newdev->paths.service_revision = strdup(path);
+}
+static void found_gatt_service(struct ble *newdev, const char *path, sd_bus_message *reply)
+{
+	bool matches = false;
+	bool service_found = false;
+	if (sd_bus_message_enter_container(reply, SD_BUS_TYPE_ARRAY, "{sv}") < 0)
+		return;
+
+	while (0 < sd_bus_message_enter_container(reply, SD_BUS_TYPE_DICT_ENTRY, "sv")) {
+		const char *prop;
+		if (0 <= sd_bus_message_read_basic(reply, 's', &prop)) {
+			if (!strcmp(prop, "Device")) {
+				const char *devpath;
+				if (0 <= sd_bus_message_read(reply, "v", "o", &devpath) &&
+				    !strcmp(devpath, newdev->paths.dev)) {
+					matches = true;
+				}
+			} else if (!strcmp(prop, "UUID")) {
+				const char *uuid;
+				if (0 <= sd_bus_message_read(reply, "v", "s", &uuid)) {
+					if (!strcmp(uuid, FIDO_SERVICE_UUID))
+						service_found = true;
+				}
+			} else {
+				sd_bus_message_skip(reply, "v");
+			}
+		}
+		sd_bus_message_exit_container(reply);
+	}
+	sd_bus_message_exit_container(reply);
+	if (matches && service_found) {
+		newdev->paths.service = strdup(path);
+	}
+}
+
+static int registered_application(sd_bus_message *reply, void *user_data, sd_bus_error *ret_error)
+{
+	(void)reply;
+	(void)ret_error;
+	*(bool *)user_data = true;
+	return 0;
+}
+
+static void disable_ble_adapters(void *data, const char *path, sd_bus_message *reply)
+{
+	const char *iface;
+	sd_bus *bus = (sd_bus *) data;
+	if (sd_bus_message_read_basic(reply, 's', &iface) >= 0) {
+		if (!strcmp(iface, DBUS_GATTMANAGER_IFACE)) {
+			bool finished = false;
+			int ret;
+			sd_bus_call_method_async(bus, NULL, "org.bluez", path, iface, "UnregisterApplication",
+						 registered_application, &finished, "o", "/org/fido");
+			while (!finished) {
+				ret = sd_bus_process(bus, NULL);
+				if (ret < 0)
+					return;
+
+				if (ret == 0) {
+					ret = sd_bus_wait(bus, UINT64_MAX);
+					if (ret <= 0)
+						return;
+				}
+			}
+		}
+		sd_bus_message_skip(reply, "a{sv}");
+	}
+}
+
+static void enable_ble_adapters(void *data, const char *path, sd_bus_message *reply)
+{
+	const char *iface;
+	struct manifest_ctx *ctx = (struct manifest_ctx *) data;
+	if (sd_bus_message_read_basic(reply, 's', &iface) >= 0) {
+		if (!strcmp(iface, DBUS_ADAPTER_IFACE)) {
+			/* will also fail if adapter is not powered */
+			if (sd_bus_call_method(ctx->bus, "org.bluez", path, iface,
+					       "StartDiscovery", NULL, NULL, "") >= 0) {
+				ctx->scanning = true;
+			}
```

Comment:
Generally speaking, we've avoided waiting for devices in `fido_dev_info_manifest()`, do we need this or can we put the burden on the desktop environment to discover to the device?

Reply:
Well, I would prefer the connection being established when the fido usage started automatically (and disconnected if not used any longer), but I understand your concern.
bluetoothctl and the bluetooth ui from KDE allow manual connects, blueman not (but still allows pairing). I would agree to remove that connection/discovery method for now, it can be added later on

Reply:
> I would prefer the connection being established when the fido usage started automatically (and disconnected if not used any longer)

Without asking for any changes at this time -- would you agree that connecting and disconnecting should in that case preferably happen in `fido_dev_open()` and `fido_dev_close()` respectively? The application may not be calling `fido_dev_info_manifest()` at all, but instead rely on a previously known device path.

>  I would agree to remove that connection/discovery method for now, it can be added later on

Indeed, given the (very) asynchronous nature of BLE, we'll probably have to think a bit on what we'd like to do. Happy to return to it at some later point.

Reply:
well, I think all the device paths look a bit wonky for all kind of devices, but having that possibility to directly open a connection should be kept.
But not sure about all that. But maybe fido_dev_open should avoid doing anything consuming more time.  Probably the better place for anything consuming more time would be in fido_dev_info_manifest().

Reply:
> But maybe fido_dev_open should avoid doing anything consuming more time.

As an aside, there are situations in which `fido_dev_open()` may block, see for example the changes introduced by https://github.com/Yubico/libfido2/pull/350.